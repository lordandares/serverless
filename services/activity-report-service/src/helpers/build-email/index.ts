import * as BPromise from 'bluebird'
import {
  filter,
  find,
  flow,
  get,
  getOr,
  isEmpty,
  reject,
  toUpper,
  trim,
} from 'lodash/fp'
import * as moment from 'moment-timezone'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

import { Location } from '../../handlers/email-processor'
import { getLocationReportLinks } from '../get-location-report-links'
import { UserDetails } from '../get-user-details'

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(true, 'options.formSubmissions'),
)

const getNoActivityOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(false, 'options.skipNoActivity'),
)

export interface BuildEmail extends UserDetails {
  locations: Location[]
}

export async function buildEmail(options: BuildEmail) {
  try {
    const {
      applicationName,
      email,
      firstName,
      locations,
      preferences,
      timezone,
    } = options

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-location-reports'
      : 'daily-location-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment()
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)
    const skipNoActivity = getNoActivityOption(preferences)

    const links = await BPromise.map(locations, location =>
      getLocationReportLinks({
        location,
        withFormSubmissions,
      }),
    )

    const activeLocations = filter('link')(links)
    const inactiveLocations = reject('link')(links)

    const hasNoLocations =
      isEmpty(activeLocations) && isEmpty(inactiveLocations)

    if (
      isEmpty(activeLocations) &&
      !isEmpty(inactiveLocations) && skipNoActivity
    ) {
      console.info('skipping activity report', {
        activeLocations: activeLocations.length,
        inactiveLocations: inactiveLocations.length,
        skipNoActivity,
      })
      return false
    }

    const name = trim(firstName)

    const data = JSON.stringify({
      activeLocations,
      applicationName,
      date: dateText,
      hasNoLocations,
      inactiveLocations,
      isProduction,
      name,
      sendInactive: !skipNoActivity,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, options })
    return false
  }
}
