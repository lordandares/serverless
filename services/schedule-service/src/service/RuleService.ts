import { errors, schemas } from '@lighthouse/serverless-common'
import { DynamoRepository } from '../repository/dynamo'
import { payloadToRulePatternDocument } from './lib/ruleTransform'

declare var process: {
  env: {
    TABLE_SCHEDULES: string
  }
}

const removeRule = async (id: string): Promise<void> => {
  if (!id) {
    throw new errors.ApplicationError({
      message: `The rule \`id\` of the resource to delete is missing`,
    })
  }

  console.debug('RemoveRule: id', { id })

  const table: string = process.env.TABLE_SCHEDULES
  const dbObject = DynamoRepository(table)
  // await dbObject.remove(id)
}

const RuleService = {
  removeRule,
}

export { RuleService }
