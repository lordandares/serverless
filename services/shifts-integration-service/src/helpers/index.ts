export { default as getPlatformEndpoint } from './get-platform-endpoint'
export { default as getSnsPayload } from './get-sns-payload'
export {
  getWinTeamMessages,
  getWinTeamTimePunch,
} from './get-winteam-time-punch'
export { default as notifyUser } from './notify-user'
export { default as parseWinTeamMessages } from './parse-winteam-messages'
export { default as sendMessage } from './send-message'
export {
  default as validateEndpointAttributes,
} from './validate-endpoint-attributes'
