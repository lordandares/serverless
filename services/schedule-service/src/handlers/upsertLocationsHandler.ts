import { errors, schemas } from '@lighthouse/serverless-common'
import * as BPromise from 'bluebird'
import { getOr } from 'lodash/fp'

import { ScheduleService } from '../service/ScheduleService'
import parseSnsEvent from './lib/parseSnsEvent'

export async function upsertLocationsHandler(
  event,
  context,
): Promise<schemas.ScheduleLocationDocumentSchema[]> {
  try {
    let payload = event

    // NOTE: this handler is invoked from a step function or an sns event
    const isSnsEvent =
      getOr(false, 'Records.0.EventSource', event) === 'aws:sns'

    if (isSnsEvent) {
      const { body } = parseSnsEvent(event)
      payload = body
    }

    if (!context) {
      throw new errors.ApplicationError({
        message: 'UpsertLocationHandler: missing context',
      })
    }

    if (!payload) {
      throw new errors.ApplicationError({
        message: 'UpsertLocationHandler: missing payload',
      })
    }

    const { applicationId, locations, scheduleId } = payload
    const { awsRequestId, invokedFunctionArn } = context

    const results: schemas.ScheduleLocationDocumentSchema[] = await BPromise.map(
      locations,
      async (
        locationId: string,
      ): Promise<schemas.ScheduleLocationDocumentSchema> => {
        const location = await ScheduleService.upsertLocation({
          applicationId,
          arn: invokedFunctionArn,
          arnId: awsRequestId,
          locationId,
          scheduleId,
        })

        return location
      },
    )

    return results
  } catch (err) {
    console.error('UpsertLocationsError', {
      err,
      event,
    })

    if (errors.isKnownError(err)) {
      throw err
    }

    throw new errors.UnknownError()
  }
}
