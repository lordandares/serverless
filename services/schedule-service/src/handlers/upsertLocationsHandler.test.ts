import { mongo, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk-mock'
import * as MockDate from 'mockdate'

import { createContext, createSnsEvent } from '../../../../__test__/helpers'
import {
  areaDocument,
  scheduleDocument,
  scheduleLocationDocument,
} from '../__fixtures__'
import { ScheduleService } from '../service/ScheduleService'
import { upsertLocationsHandler } from './upsertLocationsHandler'

beforeEach(() => {
  process.env.TABLE_SCHEDULES = 'table-schedules'
  MockDate.set('2019-10-13T01:00:00.000Z')
})

afterEach(() => MockDate.reset())

test('event body error', async () => {
  expect.assertions(1)

  const event = createSnsEvent({
    body: null,
  })

  const context = createContext()

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: UpsertLocationHandler: missing payload]`,
    )
  }
})

test('event context error', async () => {
  expect.assertions(1)

  const event = createSnsEvent({
    body: {
      applicationId: 'app1',
      locations: ['location1'],
      scheduleId: 'schedule1',
    },
  })

  const context = null

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: UpsertLocationHandler: missing context]`,
    )
  }
})

test('application error returned when missing applicationId', async () => {
  expect.assertions(1)

  const event = {
    applicationId: null,
    locations: ['location1'],
    scheduleId: 'schedule1',
  }

  const context = createContext()

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: The \`applicationId\` is missing]`,
    )
  }
})

test('application error returned when missing arn', async () => {
  expect.assertions(1)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
    scheduleId: 'schedule1',
  }

  const context = { awsRequestId: 'arnId', invokedFunctionArn: null } as any

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: The \`arn\` is missing]`,
    )
  }
})

test('application error returned when missing arnId', async () => {
  expect.assertions(1)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
    scheduleId: 'schedule1',
  }

  const context = { awsRequestId: null, invokedFunctionArn: 'arn' } as any

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: The \`arnId\` is missing]`,
    )
  }
})

test('application error returned when missing locationId', async () => {
  expect.assertions(1)

  const event = {
    applicationId: 'app1',
    locations: [null],
    scheduleId: 'schedule1',
  }

  const context = createContext()

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: The \`locationId\` is missing]`,
    )
  }
})

test('generic error', async () => {
  expect.assertions(2)

  const upsertLocationSpy = jest.spyOn(ScheduleService, 'upsertLocation')

  upsertLocationSpy.mockImplementation(
    jest.fn().mockRejectedValueOnce(new Error('Unknown Error')),
  )

  const event = createSnsEvent({
    body: {
      applicationId: 'app1',
      locations: ['location1'],
      scheduleId: 'schedule1',
    },
  })

  const context = createContext()

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(upsertLocationSpy).toHaveBeenCalled()
    expect(error).toMatchInlineSnapshot(
      `[UnknownError: Something went wrong! Try again or contact support if the problem persists.]`,
    )
  }

  upsertLocationSpy.mockRestore()
})

test('application error when validation error', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: undefined })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const validateSpy = jest.spyOn(schemas, 'validate')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))
  validateSpy.mockImplementation(
    jest.fn().mockRejectedValueOnce(new Error('Location has malformed data')),
  )

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  const event = createSnsEvent({
    body: {
      applicationId: 'app1',
      locations: ['location1'],
      scheduleId: 'schedule1',
    },
  })

  const context = createContext()

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: Location has malformed data]`,
    )
  }

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()
  validateSpy.mockRestore()

  AWS.restore()
})

test('location error', async () => {
  expect.assertions(1)

  const getSpy = jest
    .fn()
    .mockRejectedValue(new Error('Error getting location'))
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  const event = createSnsEvent({
    body: {
      applicationId: 'app1',
      locations: ['location1'],
      scheduleId: 'schedule1',
    },
  })

  const context = createContext()

  try {
    await upsertLocationsHandler(event, context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ApplicationError: Error getting location]`,
    )
  }

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('returns created document from sns event', async () => {
  expect.assertions(2)

  const getSpy = jest.fn().mockResolvedValue({ Item: undefined })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const putSpy = jest.fn().mockResolvedValue({})
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)

  const event = createSnsEvent({
    body: {
      applicationId: 'app1',
      locations: ['location1'],
      scheduleId: 'schedule1',
    },
  })

  const context = createContext()

  const result = await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()
  expect(result).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('returns created document from simple payload', async () => {
  expect.assertions(2)

  const getSpy = jest.fn().mockResolvedValue({ Item: undefined })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const putSpy = jest.fn().mockResolvedValue({})
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
    scheduleId: 'schedule1',
  }

  const context = createContext()

  const result = await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()
  expect(result).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('syncs schedules when given associated location schedules and schedule id', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: scheduleLocationDocument })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const querySpy = jest.fn().mockResolvedValue({ Items: [scheduleDocument] })

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
    scheduleId: 'schedule1',
  }

  const context = createContext()

  await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('syncs schedules when given associated location schedules and no schedule id', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: scheduleLocationDocument })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const querySpy = jest.fn().mockResolvedValue({ Items: [scheduleDocument] })

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
  }

  const context = createContext()

  const result = await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('syncs schedules when no given associated location schedules and schedule id', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: scheduleLocationDocument })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const querySpy = jest.fn().mockResolvedValue({ Items: [] })

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
    scheduleId: 'schedule1',
  }

  const context = createContext()

  const result = await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('syncs schedules when no given associated location schedules and no schedule id', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: scheduleLocationDocument })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const querySpy = jest.fn().mockResolvedValue({ Items: [] })

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
  }

  const context = createContext()

  const result = await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('doesnt sync any schedules when no location and no schedule id', async () => {
  expect.assertions(1)

  const getSpy = jest.fn().mockResolvedValue({ Item: undefined })
  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const querySpy = jest.fn().mockResolvedValue({ Items: [] })

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('DynamoDB.DocumentClient', 'query', querySpy)

  const event = {
    applicationId: 'app1',
    locations: ['location1'],
  }

  const context = createContext()

  const result = await upsertLocationsHandler(event, context)

  expect(putSpy.mock.calls).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})
