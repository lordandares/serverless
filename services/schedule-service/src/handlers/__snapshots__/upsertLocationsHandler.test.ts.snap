// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesnt sync any schedules when no location and no schedule id 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T01:00:00.000Z",
        "createdBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns created document from simple payload 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T01:00:00.000Z",
        "createdBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [
          "schedule1",
        ],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns created document from simple payload 2`] = `
Array [
  Object {
    "applicationId": "app1",
    "createdAt": "2019-10-13T01:00:00.000Z",
    "createdBy": Object {
      "id": "awsRequestId",
      "label": "invokedFunctionArn",
      "type": "system",
    },
    "data": Object {
      "name": "WRNMMC Facility",
      "serviceHours": Object {
        "hours": Array [
          Object {
            "description": "MON 00:00 - TUE 00:00",
            "duration": 86400000,
            "end": 86400000,
            "start": 0,
            "type": "DEFAULT",
          },
        ],
        "timezone": "America/New_York",
      },
    },
    "groupType": "location",
    "pk": "app1-location",
    "schedules": Array [
      "schedule1",
    ],
    "sk": "location1",
    "updatedAt": "2019-10-13T01:00:00.000Z",
    "updatedBy": Object {
      "id": "awsRequestId",
      "label": "invokedFunctionArn",
      "type": "system",
    },
  },
]
`;

exports[`returns created document from sns event 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T01:00:00.000Z",
        "createdBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [
          "schedule1",
        ],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns created document from sns event 2`] = `
Array [
  Object {
    "applicationId": "app1",
    "createdAt": "2019-10-13T01:00:00.000Z",
    "createdBy": Object {
      "id": "awsRequestId",
      "label": "invokedFunctionArn",
      "type": "system",
    },
    "data": Object {
      "name": "WRNMMC Facility",
      "serviceHours": Object {
        "hours": Array [
          Object {
            "description": "MON 00:00 - TUE 00:00",
            "duration": 86400000,
            "end": 86400000,
            "start": 0,
            "type": "DEFAULT",
          },
        ],
        "timezone": "America/New_York",
      },
    },
    "groupType": "location",
    "pk": "app1-location",
    "schedules": Array [
      "schedule1",
    ],
    "sk": "location1",
    "updatedAt": "2019-10-13T01:00:00.000Z",
    "updatedBy": Object {
      "id": "awsRequestId",
      "label": "invokedFunctionArn",
      "type": "system",
    },
  },
]
`;

exports[`syncs schedules when given associated location schedules and no schedule id 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "123456789",
          "label": "label",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [
          "schedule1",
        ],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
  Array [
    Object {
      "Item": Object {
        "applicationId": "application1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "user1",
          "label": "Unknown User",
          "type": "user",
        },
        "data": Object {
          "areas": Array [
            "area1",
          ],
          "enabled": true,
          "included": Object {
            "locations": Object {
              "location1": Object {
                "name": "WRNMMC Facility",
              },
            },
          },
          "locations": Array [
            "location1",
          ],
          "name": "Schedule One",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 11:00 - TUE 00:00",
                "end": 86400000,
                "start": 39600000,
                "type": "DEFAULT",
              },
            ],
            "timezone": "Australia/Melbourne",
          },
          "strategy": Object {
            "options": Object {
              "duration": Object {
                "unit": "minute",
                "value": 15,
              },
              "frequency": Object {
                "unit": "minute",
                "value": 15,
              },
            },
            "type": "stopwatch",
          },
          "type": "visit",
        },
        "endAt": "2019-10-1500:00:00.000Z",
        "groupType": "schedule",
        "pk": "application1-schedule",
        "scheduleId": "schedule1",
        "sk": "2019-10-13T00:00:00.000Z-schedule1",
        "startAt": "2019-10-1400:00:00.000Z",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "user1",
          "label": "Unknown User",
          "type": "user",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`syncs schedules when given associated location schedules and schedule id 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "123456789",
          "label": "label",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [
          "schedule1",
        ],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
  Array [
    Object {
      "Item": Object {
        "applicationId": "application1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "user1",
          "label": "Unknown User",
          "type": "user",
        },
        "data": Object {
          "areas": Array [
            "area1",
          ],
          "enabled": true,
          "included": Object {
            "locations": Object {
              "location1": Object {
                "name": "WRNMMC Facility",
              },
            },
          },
          "locations": Array [
            "location1",
          ],
          "name": "Schedule One",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 11:00 - TUE 00:00",
                "end": 86400000,
                "start": 39600000,
                "type": "DEFAULT",
              },
            ],
            "timezone": "Australia/Melbourne",
          },
          "strategy": Object {
            "options": Object {
              "duration": Object {
                "unit": "minute",
                "value": 15,
              },
              "frequency": Object {
                "unit": "minute",
                "value": 15,
              },
            },
            "type": "stopwatch",
          },
          "type": "visit",
        },
        "endAt": "2019-10-1500:00:00.000Z",
        "groupType": "schedule",
        "pk": "application1-schedule",
        "scheduleId": "schedule1",
        "sk": "2019-10-13T00:00:00.000Z-schedule1",
        "startAt": "2019-10-1400:00:00.000Z",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "user1",
          "label": "Unknown User",
          "type": "user",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`syncs schedules when no given associated location schedules and no schedule id 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "123456789",
          "label": "label",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [
          "schedule1",
        ],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`syncs schedules when no given associated location schedules and schedule id 1`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "app1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "123456789",
          "label": "label",
          "type": "system",
        },
        "data": Object {
          "name": "WRNMMC Facility",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 00:00 - TUE 00:00",
                "duration": 86400000,
                "end": 86400000,
                "start": 0,
                "type": "DEFAULT",
              },
            ],
            "timezone": "America/New_York",
          },
        },
        "groupType": "location",
        "pk": "app1-location",
        "schedules": Array [
          "schedule1",
        ],
        "sk": "location1",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "awsRequestId",
          "label": "invokedFunctionArn",
          "type": "system",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;
